src/cascadia/TerminalSettingsModel/IconPathConverter.cpp

```diff
@@ -197,13 +209,175 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+        if (!til::ends_with(iconPathWithoutIndex, L".exe") &&
+            !til::ends_with(iconPathWithoutIndex, L".dll") &&
+            !til::ends_with(iconPathWithoutIndex, L".lnk"))
+        {
+            return std::nullopt;
+        }
+
+        if (commaIndex != std::wstring::npos)
+        {
+            // Convert the string iconIndex to an int
+            const auto index = til::to_ulong(pathView.substr(commaIndex + 1));
+            if (index == til::to_ulong_error)
+            {
+                return std::nullopt;
+            }
+            return static_cast<int>(index);
+        }
+
+        // We had a binary path, but no index. Default to 0.
+        return 0;
+    }
+
+    winrt::Windows::Foundation::IAsyncOperation<winrt::Windows::UI::Xaml::Media::Imaging::SoftwareBitmapSource> _getImageIconSourceForBinary(std::wstring_view iconPathWithoutIndex,
+                                                                                                                                             int index)
+    {
+        winrt::apartment_context fg_thread;
+
+        // Try:
+        // * c:\Windows\System32\SHELL32.dll, 210
+        // * c:\Windows\System32\notepad.exe, 0
+        // * C:\Program Files\PowerShell\6-preview\pwsh.exe, 0 (this doesn't exist for me)
+        // * C:\Program Files\PowerShell\7\pwsh.exe, 0
+
+        co_await winrt::resume_background();
+
+        auto swBitmap{ _getBitmapFromIconFileAsync(winrt::hstring{ iconPathWithoutIndex }, index, 32) };
+        if (swBitmap == nullptr)
+        {
+            co_return nullptr;
+        }
+
+        co_await fg_thread;
+        winrt::Windows::UI::Xaml::Media::Imaging::SoftwareBitmapSource bitmapSource{};
+        co_await bitmapSource.SetBitmapAsync(swBitmap);
+        co_await fg_thread;
+        co_return bitmapSource;
+    }
+
+    winrt::Windows::UI::Xaml::Media::Imaging::SoftwareBitmapSource _getImageIconSourceForBinaryNotAsync(std::wstring_view iconPathWithoutIndex,
+                                                                                                        int index)
+    {
+        auto swBitmap{ _getBitmapFromIconFileAsync(winrt::hstring{ iconPathWithoutIndex }, index, 32) };
+        if (swBitmap == nullptr)
+        {
+            return nullptr;
+        }
+
+        winrt::Windows::UI::Xaml::Media::Imaging::SoftwareBitmapSource bitmapSource{};
+        bitmapSource.SetBitmapAsync(swBitmap);
+        return bitmapSource;
+    }
+
+    MUX::Controls::IconSource IconPathConverter::IconSourceMUX(const winrt::hstring& iconPath)
+    {
+        std::wstring_view iconPathWithoutIndex;
+        auto indexOpt = _getIconIndex(iconPath, iconPathWithoutIndex);
+        if (!indexOpt.has_value())
+        {
+            return _IconSourceMUX(iconPath);
+        }
+        auto index = indexOpt.value();
+
+        auto bitmapSource = _getImageIconSourceForBinaryNotAsync(iconPathWithoutIndex, index);

```

Comment:
```suggestion
        const auto bitmapSource = _getImageIconSourceForBinaryNotAsync(iconPathWithoutIndex, indexOpt.value());
```
nit: you can get rid of index or add const to both